### ISSUE TITLE
 Warn when detecting libraries and other unreadable code

### ISSUE THREAD

**Issue by diox** [mozilla/addons-linter#1323]
For post-review, we want to increase priority of add-ons using unknown libraries we haven't specifically allowed (or denied) or other unreadable code. The linter should detect this and add a warning with a specific code that addons-server can look for.

This is vague, but I think a basic obfuscation detection is what we want here. @jvillalobos and @wagnerand for more details.

**Comment by wagnerand** [comment]
Yes, obfuscation is what comes into my mind here. I *think* that should cover it pretty much, but let's hear from @jvillalobos to see whether there is something beyond that.

**Comment by jvillalobos** [comment]
Yes, we want to detect code that is either minified or obfuscated. @kewisch pointed out that whitespace density could be a good indicator of this.

**Comment by kewisch** [comment]
Also, complete lack of code comments (although that should not be a factor alone)

**Comment by jvillalobos** [comment]
Somewhat related, in a meeting today @cr brought up retire.js (https://retirejs.github.io/retire.js/) as a tool that can detect libraries and flag outdated ones. I don't know if we're considered this for the linter before, but it looks like an interesting option.

**Comment by EnTeQuAk** [comment]
There's already an issue for that, see https://github.com/mozilla/addons-linter/issues/699 for a few more details.

**Comment by EnTeQuAk** [comment]
@diox @wagnerand I wonder if this should be a real warning for other users too or just a flag in the metadata so that addons-server knows about this given that we allow uploading of sources and stuff on AMO?

Maybe a warning that says "please remember to upload sources if you submit minified or otherwise unreadable code"?

**Comment by wagnerand** [comment]
> Maybe a warning that says "please remember to upload sources if you submit minified or otherwise unreadable code"?

We are doing that already with mixed results.

I think the main idea here is to have a risk factor that detects minified code.

**Comment by EnTeQuAk** [comment]
> I think the main idea here is to have a risk factor that detects minified code.

which means it'll just be added to the metadata similar to how jsLibs or other flagged files are. That sounds like a plan.

**Comment by EnTeQuAk** [comment]
Implementation / Experimentation path:

**Easier, approach matching everything potentially minified:**
 * Does the file end with ".min.js"? (adds to a `could_be_minified` score)

From firefox devtools (also adds to a `could_be_minified` score):

```javascript 
const SAMPLE_SIZE = 30; // no of lines
const INDENT_COUNT_THRESHOLD = 20; // percentage

/**
 * Determines is the source text is minified.
 * Using the percentage no. of the indented lines from a sample set of lines to determine if the js file
 * is minified
 * @return boolean
 *          True if source text is minified.
 */
isMinified: function(aText){
  let lineEndIndex = 0;
  let lineStartIndex = 0;
  let lines = 1;
  let indentCount = 0;
  // Strip comments.
  aText = aText.replace(/\/\*[\S\s]*?\*\/|\/\/.+/g, "");
  while(lines < SAMPLE_SIZE){
    lineEndIndex = aText.indexOf("\n", lineStartIndex);
    if(lineEndIndex == -1){
       break;
    }
    if(/^\s+/.test(aText.slice(lineStartIndex, lineEndIndex))){
      indentCount++;
    }
    lineStartIndex = lineEndIndex + 1;
    lines++;
  }
  return (indentCount * 100) < INDENT_COUNT_THRESHOLD;
}
```

which seems to be fairly good at what it does. WebKit in addition matches lines that are longer than 500 characters which sounds like a good idea but also super-naive and broad.

*Medium time intensive:*

Write parser that searches for code based on https://github.com/RetireJS/retire.js/blob/master/repository/jsrepository.json - see `exractors`, e.g https://github.com/RetireJS/retire.js/blob/master/repository/jsrepository.json#L64-L84 and matches libraries more broadly than dispensary does. This integration could actually end up being an improvement to dispensary detecting libraries and their version numbers and white/blacklisting them.

These extractors can improve our `can_be_minified` score or make it less brittle.

*Hard, time intensive:*

 * find / read package.json
 * find / read npm-shrinkwrap.json
 * find / read yarn.lock
 * find / read package-lock.json
 * Try to figure out common data-structures and build a dependency-map which we can then use to warn accordingly

Question here is, are we concerned about things that are already, potentially properly managed via package.json? Not even sure if these files are in the XPIs we see from day-to-day?

So, my action plan right now would be:

 * Start with the easier approach, try to see how far that detection goes on a sample of 250 WebExtensions
 * If that results in being too broadly, try the medium time intensive approach and use Retire.js repository rules for matching
 * I'm sure we'll end up with something reasonable at this point that is an improvement already, if not, go hardcode and parse packages from metadata and go all bonkers

**Comment by muffinresearch** [comment]
> 500 characters which sounds like a good idea but also super-naive and broad.

Actually that seems quite reasonable. Un-minified code is very unlikely to hit that limit.

**Comment by muffinresearch** [comment]
@cgrebs this all looks really good to me. Definitely starting with the easier stuff is a very good idea.

**Comment by EnTeQuAk** [comment]
> Actually that seems quite reasonable. Un-minified code is very unlikely to hit that limit.

Well, if I paste some base64 string or some copied list from somewhere I'm sure this might happen. But yeah, it's still a good enough guess if weighted correctly.

**Comment by EnTeQuAk** [comment]
I mark this as fixed for now since we have the detection, I'll add optional issues for later that are ideas for future improvements but if the current implementation works we might not need them.

It'd be nice to have retire.js integration for other reasons though so this should stay somewhat in focus.

**Comment by ValentinaPC** [comment]
@EnTeQuAk : Could you provide one or two examples with unknown libraries or other unreadable code? Thanks in advance!

**Comment by EnTeQuAk** [comment]
@ValentinaPC you can theoretically browse through https://cdnjs.com/libraries and use anything that is `.min.js` and not in https://github.com/mozilla/dispensary/blob/master/src/libraries.json (which is the list of libraries we do know about). Afaik the risk factor increase for unknown minified code is already on -dev (or even prod? I don't remember) so you should see it working properly there. Remember, the score get's increased only once by 10 completely unrelated how many libraries the add-on uses.

**Comment by ValentinaPC** [comment]
@EnTeQuAk : Hey, Chris, any news on this? Thanks!

**Comment by EnTeQuAk** [comment]
I attached a zip that contains a jQuery 3.2.1 that is modified to not throw any other warnings so you should now be able to test that.

[jquery-modified.zip](https://github.com/mozilla/addons-linter/files/1333815/jquery-modified.zip)

**Comment by ValentinaPC** [comment]
Verified on AMO-dev FF55(Win 7).
While adding the above .zip to the suited add-on the following warning was threw:
![d](https://user-images.githubusercontent.com/15685960/30917097-ca765430-a3a3-11e7-96af-fb45bc5d6833.png)
Weight was incremented with 10.
I don't know if I did it right?
Add-on: https://addons-dev.allizom.org/en-US/firefox/addon/webext-libun/

**Comment by EnTeQuAk** [comment]
>  Weight was incremented with 10.

That sounds perfect!
